design/DESIGN_typed_context_pipeline.md

```diff
@@ -0,0 +1,451 @@
+      .build(mySecureHandler);                   // Handler expects UserSessionContextView - OK!
+
+  // This handler can now be passed to Relic's server serving mechanism.
+  // relicServe(requestHandler, ...);
+
+  // Example of a compile-time error:
+  // final faultyHandler = PipelineBuilder.start()
+  //     .add(authenticationMiddleware) // Output: UserContextView
+  //     // Missing sessionMiddleware
+  //     .build(mySecureHandler);      // COMPILE ERROR: mySecureHandler expects UserSessionContextView,
+  //                                  // but pipeline only guarantees UserContextView.
+}
+```
+
+## 6. Benefits
+
+*   **Compile-Time Safety**: The primary goal. Misconfigured pipelines (e.g., missing middleware, incorrect order affecting context data) are caught by the Dart analyzer.
+*   **Improved Developer Ergonomics**:
+    *   Handlers and middleware can declare precisely the context view (and thus data) they expect.
+    *   Access to context data via extension type getters is type-safe and clear (e.g., `context.user`).
+*   **Minimal Runtime Overhead for Views**: Extension types are intended to be zero-cost compile-time wrappers. The `ContextProperty` helper encapsulates `Expando` lookups/attachments, which are generally efficient.
+*   **No Class Explosion**: Avoids needing a distinct context `class` for every possible combination of middleware. Extension types provide views, and `ContextProperty` manages data association with the stable `requestToken`.
+*   **Clarity and Documentation**: The type signatures of middleware and handlers explicitly document their context dependencies. View methods (e.g., `attachUser`) and `ContextProperty` provide clear, discoverable APIs for data management.
+*   **Modularity & Encapsulation**: `ContextProperty` encapsulates `Expando` usage. Modules define their data properties cleanly.
+
+## 7. Middleware Paradigm Shift and Implications
+
+This typed pipeline introduces a shift from the traditional `Middleware = Handler Function(Handler innerHandler)` pattern previously used. Understanding these changes is crucial:
+
+*   **New Middleware Signature**: Middleware in this design are functions with a signature like `OutputView Function(InputView)`. They transform context views rather than wrapping an inner handler.
+*   **Linear Chain**: The `PipelineBuilder` composes middleware into a linear chain of context transformations. Each middleware is expected to process the context and pass control (via its return value) to the next stage defined by the builder.
+*   **Short-Circuiting (e.g., Denying Access)**:
+    *   Middleware should not directly return a `Response` to short-circuit the pipeline.
+    *   Instead, if a middleware needs to stop processing and return an error (e.g., an authentication middleware denying access due to an invalid token), it should **throw a specific exception** (e.g., `AuthorizationRequiredError("Invalid token")`, `PaymentRequiredError()`).
```

Comment:
Tricky, personally I was not a fan of the `throw`.

But maybe this could be handle with some `Result`-ish type which distinct between "early return" and "next".

But either way, I think that is a secondary consideration and would not need to happen in combination with the major change of this proposal, right?

Reply:
Yes, having just fought to get rid of `HijackException` based control flow I know where you are coming from.

However, you won't inadvertently catch these exceptions in the pipeline as the middleware don't wrap each other as today.

I kind of made peace with it. Passing a context to a handler that it shouldn't react to is also a bit weird, and then there is the question of typing.